src/libraries/System.Private.CoreLib/src/System/Threading/Wasi/WasiEventLoop.cs

```diff
@@ -4,72 +4,120 @@
 using System.Collections.Generic;
 using System.Threading.Tasks;
 using WasiPollWorld.wit.imports.wasi.io.v0_2_1;
+using Pollable = WasiPollWorld.wit.imports.wasi.io.v0_2_1.IPoll.Pollable;
 
 namespace System.Threading
 {
     internal static class WasiEventLoop
     {
-        private static List<WeakReference<TaskCompletionSource>> s_pollables = new();
+        // TODO: if the Pollable never resolves and and the Task is abandoned
```

Comment:
[Here's my naive attempt at this](https://github.com/dotnet/runtimelab/pull/2614/files#diff-05dfa88296e46caaf99839b1df9539404e352ee7a3b06c13ea14b5dbda83d917).  I'm sure yours is better, but one thing to make sure of is that we give all tasks a chance to run once any of them have been cancelled (see [this comment](https://github.com/dotnet/runtimelab/pull/2614/files#diff-05dfa88296e46caaf99839b1df9539404e352ee7a3b06c13ea14b5dbda83d917R67-R70).  I can't tell if your code accounts for this.

FWIW, [this](https://github.com/dotnet/runtimelab/pull/2614/files#diff-67fd247f21f96df8bb587b0caf914dcb65cbf1e941d9e2e2ab5f54ddfa8a5740R176-R185) is the code I've been using to test `HttpClient` timeouts.  It's paired with [this server code](https://github.com/dotnet/runtimelab/pull/2614/files#diff-add4766038a7ab88d051e74962bdb93b54737142b17521136365cfb61adb790bR75-R84).

Reply:
My solution will cancel the `Task` synchronously when the `CancellationTokenSource.Cancel` is called because of the `CancellationTokenRegistration` callback. 
And that could possibly trigger execution of other synchronous/inline continuations.
It seems to me, that synchronous cancellation is  better. And it's enough to pass the test.

I added test similar to yours with some azure endpoint, which is not perfect, but I plan to fix it later.